From 72be9a408eeb1f5c59b4001ec27093147022a0a0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pasi=20Sj=C3=B6holm?= <pasi.sjoholm@jollamobile.com>
Date: Mon, 23 Mar 2015 09:38:03 +0200
Subject: [PATCH 17/22] service: Set routes only for same address family

Routes should not be tried to be set with nameserver and gateway
being of different address family, as this will lead into having
link-local routes for hosts which are not on the link
(add_nameserver_route() will add link-local route if adding host
route fails).

(cherry picked from commit 60d5afa75e798f78374081a9d6b4d86c534d8ab6)
---
 src/service.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/service.c b/src/service.c
index e797e7e..332ffd4 100644
--- a/src/service.c
+++ b/src/service.c
@@ -1295,14 +1295,18 @@ static void add_nameserver_route(int family, int index, char *nameserver,
 static void nameserver_add_routes(int index, char **nameservers,
 					const char *gw)
 {
-	int i, family;
+	int i, ns_family, gw_family;
+
+	gw_family = connman_inet_check_ipaddress(gw);
+	if (gw_family < 0)
+		return;
 
 	for (i = 0; nameservers[i]; i++) {
-		family = connman_inet_check_ipaddress(nameservers[i]);
-		if (family < 0)
+		ns_family = connman_inet_check_ipaddress(nameservers[i]);
+		if (ns_family < 0 || ns_family != gw_family)
 			continue;
 
-		add_nameserver_route(family, index, nameservers[i], gw);
+		add_nameserver_route(ns_family, index, nameservers[i], gw);
 	}
 }
 
-- 
1.9.1

